src/Scout/ScoutEngine.php

```diff
@@ -0,0 +1,542 @@
+
+        $objectIds = array_column($results, '_id');
+        $objectIdPositions = array_flip($objectIds);
+
+        return $model->queryScoutModelsByIds(
+            $builder,
+            $objectIds,
+        )->cursor()->filter(function ($model) use ($objectIds) {
+            return in_array($model->getScoutKey(), $objectIds);
+        })->map(function ($model) use ($results, $objectIdPositions) {
+            $result = $results[$objectIdPositions[$model->getScoutKey()]] ?? [];
+
+            foreach ($result as $key => $value) {
+                if (substr($key, 0, 1) === '_') {
+                    $model->withScoutMetadata($key, $value);
+                }
+            }
+
+            return $model;
+        })->sortBy(function ($model) use ($objectIdPositions) {
+            return $objectIdPositions[$model->getScoutKey()];
+        })->values();
+    }
+
+    /**
+     * Create the MongoDB Atlas Search index.
+     *
+     * Accepted options:
+     *  - wait: bool, default true. Wait for the index to be created.
+     *
+     * @see Engine::createIndex()
+     *
+     * @param string            $name    Collection name
+     * @param array{wait?:bool} $options
+     */
+    #[Override]
+    public function createIndex($name, array $options = []): void
+    {
+        assert(is_string($name), new TypeError(sprintf('Argument #1 ($name) must be of type string, %s given', get_debug_type($name))));
+
+        // Ensure the collection exists before creating the search index
+        $this->database->createCollection($name);
+
+        $collection = $this->database->selectCollection($name);
+        $collection->createSearchIndex(
+            self::DEFAULT_DEFINITION,
+            ['name' => self::INDEX_NAME],
+        );
+
+        if ($options['wait'] ?? true) {
+            $this->wait(function () use ($collection) {
+                $indexes = $collection->listSearchIndexes([
+                    'name' => self::INDEX_NAME,
+                    'typeMap' => ['root' => 'bson'],
+                ]);
+
+                return $indexes->current() && $indexes->current()->status === 'READY';
+            });
+        }
+    }
+
+    /**
+     * Delete a "search index", i.e. a MongoDB collection.
+     *
+     * @see Engine::deleteIndex()
+     */
+    #[Override]
+    public function deleteIndex($name): void
+    {
+        assert(is_string($name), new TypeError(sprintf('Argument #1 ($name) must be of type string, %s given', get_debug_type($name))));
+
+        $this->database->selectCollection($name)->drop();
+    }
+
+    /** Get the MongoDB collection used to search for the provided model */
+    private function getSearchableCollection(Model|EloquentCollection $model): MongoDBCollection
+    {
+        if ($model instanceof EloquentCollection) {
+            $model = $model->first();
+        }
+
+        assert(method_exists($model, 'searchableAs'), sprintf('Model "%s" must use "%s" trait', $model::class, Searchable::class));
```

Comment:
Is there any reason you didn't do similar assertions elsewhere before assuming a model used the Searchable trait?

Reply:
Every method is covered by this check. I check `searchableAs` in `getSearchableCollection` and `indexableAs` in `getIndexableCollection`. Now I remember this is why I instanciated the `$collection` early in update (see [your comment](https://github.com/mongodb/laravel-mongodb/pull/3205#discussion_r1859820934)). But now I check `toSearchableArray` in update.